cpp/include/cuspatial/experimental/iterator_factory.cuh

```diff
@@ -224,6 +260,112 @@ auto make_vec_2d_output_iterator(Iter d_points_begin)
+ * @param min_y Iterator to beginning of `y` data for first box vertices.
+ * @param max_x Iterator to beginning of `x` data for second box vertices.
+ * @param max_y Iterator to beginning of `y` data for second box vertices.
+ * @return Iterator to `box`
+ *
+ * @pre Input iterators must iterate on the same data type.
+ *
+ * [LinkLRAI]: https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator
+ * "LegacyRandomAccessIterator"
+ */
+template <typename MinXIter, typename MinYIter, typename MaxXIter, typename MaxYIter>
+auto make_box_output_iterator(MinXIter min_x, MinYIter min_y, MaxXIter max_x, MaxYIter max_y)
+{
+  using T         = typename std::iterator_traits<MinXIter>::value_type;
+  auto zipped_out = thrust::make_zip_iterator(min_x, min_y, max_x, max_y);
+  return thrust::transform_output_iterator(zipped_out, detail::box_to_tuple<T>());
+}
+
+/**
+ * @brief Create an input iterator that generates zero-based sequential geometry IDs for each
+ * element based on the input offset range.
+ *
+ * This can be used for any single-level geometry offsets, e.g. multipoints, multilinestrings,
+ * (multi)trajectories. And using custom iterators it can be used for nested types.
+ *
+ * Example:
+ * @code
+ * auto offsets    = std::vector<int>({0, 3, 5, 9});
+ * auto iter_first = make_geometry_id_iterator<int>(offsets.begin(), offsets.end());
+ * auto ids = std::vector<int>(10);
+ * std::copy_n(iter_first, 10, ids.begin());  // ids now contains [0, 0, 0, 1, 1, 2, 2, 2, 2, 3]
+ * @endcode
+ *
+ * @tparam GeometryIter The offset iterator type. Must meet the requirements
+ * of [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible.
+ * @param offsets_begin Beginning of range of geometry offsets
+ * @param offsets_end End of range of geometry offsets
+ * @return An iterator over unique IDs for each element of each geometry defined by the offsets
+ *
+ * [LinkLRAI]: https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator
+ * "LegacyRandomAccessIterator"
+ */
+template <typename IndexT, typename GeometryIter>
+auto make_geometry_id_iterator(GeometryIter offsets_begin, GeometryIter offsets_end)
+{
+  return detail::make_counting_transform_iterator(
+    IndexT{0}, detail::index_to_geometry_id<IndexT, GeometryIter>{offsets_begin, offsets_end});
+}
+
+/**
+ * @brief Create an input iterator that generates zero-based sequential geometry IDs for each
+ * element of a nested geometry based on the input geometry and part offset ranges.
+ *
+ * This can be used for any two-level nested multigeometry offsets, e.g. multipolygons.
+ *
+ * Example:
+ * @code
+ * auto poly_offsets = std::vector<int>({0, 1, 3});  // poly 1 has 2 rings
+ * auto ring_offsets = std::vector<int>({0, 4, 7, 10});
+ * auto iter =
+ *   make_geometry_id_iterator<int>(poly_offsets.begin(), poly_offsets.end(), ring_offsets.begin());
+ * auto ids = std::vector<int>(13);
+ * std::copy_n(iter, 13, ids.begin());
```

Comment:
Curious where you're getting the hard coded `13` from, I think this should be the `n+1` last index so that ring offsets would be `[0, 4, 7, 10, 13` if it were in Arrow offset format?

Reply:
In that case there will be 4 rings, the poly_offsets should also be {0, 1, 3, 4}.